src/app/modules/submission/encrypt-submission/encrypt-submission.controller.ts

```diff
@@ -976,3 +981,52 @@ export const handleGetMetadata = [
   }),
   getMetadata,
 ] as ControllerHandler[]
+
+const getS3PresignedPostData: ControllerHandler<
+  unknown,
+  AttachmentPresignedPostDataMapType[] | ErrorDto,
+  AttachmentSizeMapType[]
+> = async (req, res) => {
+  const logMeta = {
+    action: 'handleGetS3PresignedPostData',
+    ...createReqMeta(req),
+  }
+  return (await getQuarantinePresignedPostData(req.body))
+    .map((presignedUrls) => {
+      logger.info({
+        message: 'Successfully retrieved quarantine presigned post data.',
+        meta: logMeta,
+      })
+      return res.json(presignedUrls)
+    })
+    .mapErr((error) => {
+      logger.error({
+        message: 'Failure getting quarantine presigned post data.',
+        meta: logMeta,
+        error,
+      })
+      const { statusCode, errorMessage } = mapRouteError(error)
+      return res.status(statusCode).json({
+        message: errorMessage,
+      })
+    })
+}
+
+// Handler for POST /:formId/submissions/storage/get-s3-presigned-post-data
+export const handleGetS3PresignedPostData = [
+  celebrate({
+    [Segments.BODY]: Joi.array()
+      .items(
+        Joi.object().keys({
+          id: Joi.string()
+            .regex(/^[0-9a-fA-F]{24}$/) // IDs should be MongoDB ObjectIDs
+            .required(),
+          size: Joi.number()
+            .max(fileSizeLimitBytes(FormResponseMode.Encrypt)) // Max attachment size is 20MB
+            .required(),
+        }),
+      )
+      .unique('id'), // IDs of each array item should be unique
```

Comment:
you _could_ also use `.custom(...)` here to check that the file size does not exceed 20MB in total :sweat_smile:

Reply:
Ooh interesting 🤔 Okay that'll probably be cleaner, let's do that instead

Reply:
Or rather, I'll add that so our code is more defensive 👍

Reply:
Thanks! Done in [00176fa](https://github.com/opengovsg/FormSG/pull/6685/commits/00176fa682be713828d2e55c2214bcf4633eee6f)